### ISSUE TITLE
Spurious errors using Sidekiq.redis within custom middleware

### ISSUE THREAD

**Issue by marclennox** [sidekiq/sidekiq#5869]
Question about using `Sidekiq.redis` within custom Sidekiq middleware.

I have a custom middleware that has been working fine on Sidekiq 7.x and Redis Client < 0.14.0, which uses `Sidekiq.redis` in order to store and retrieve temporary values in Redis related to jobs.

There was a problem with Redis Client 0.14.0 that prevented me from upgrading from Redis Client 0.13.x, but I believe that is fixed now.

In trying to upgrade now to Sidekiq 7.0.8 and Redis Client 0.14.1, I'm seeing spurious errors being raised from my custom middleware when making a Redis call using `Sidekiq.redis`, as follows:

```
RedisClient::ConnectionError: stream closed in another thread
```

Question I have is whether using `Sidekiq.redis` within middleware is OK, and if so, has something changed recently in terms of how this should be used?

Thanks.

**Comment by mperham** [comment]
As of Sidekiq 6.5, middleware should simply use `redis` and not `Sidekiq.redis`. Why? It will use a Redis connection associated with the current capsule for the running job. The global redis call will always use a connection from the global "internal" pool, which is limited in size and could be in heavier contention.

tl;dr: `Sidekiq.redis` will work but is not optimal. `redis` is optimal.

https://github.com/sidekiq/sidekiq/blob/main/docs/middleware.md#updated-api


Now that error sounds like a thread-safety issue where two threads are accessing the same connection at the same time. Are you checking out a connection and passing it around?

**Comment by marclennox** [comment]
Thanks @mperham, and I will change from `Sidekiq.redis` to `redis` as you have suggested.

To answer your question regarding the error, I don't pass around the connection per se, but I do have a PORO class which does all the "work" of interacting with Redis.  This keeps the middleware definition very small and reduces code repetition.

It is inside this PORO that I'm using `Sidekiq.redis`

Would it be safe for me to pass the redis connection from the middleware definition into the PORO, as in the following

```
redis_connection = redis { |c| c }
poro = Poro.new(c)
```

**Comment by mperham** [comment]
You cannot safely use “c” outside of the block. That’s very very bad.

**Comment by marclennox** [comment]
Fair enough.  I'll pass in the context to the PORO and make individual calls with the block format.

Strangely, this was working fine prior to redis client 0.14.0, with a very high volume of parallel workers.

I noticed that for client middleware, there is a 4th argument passed in, `redis_pool`.  When should this be used over the `redis` block approach?

**Comment by mperham** [comment]
Yeah, in order to support sharding client-side middleware passes in the pool for the relevant Redis. Every Sidekiq process only ever works against one Redis so server-side middleware only needs `redis {|c| PORO.new(c) }`. Your client-side middleware should do `redis_pool.with {|c| PORO.new(c).run }`.

**Comment by mperham** [comment]
I'm realizing that sharding may make [this comment](https://github.com/sidekiq/sidekiq/blob/934ccd257005648f88914161c8755cc5bbd19964/lib/sidekiq/middleware/modules.rb#L19) incorrect.

**Comment by marclennox** [comment]
> Yeah, in order to support sharding client-side middleware passes in the pool for the relevant Redis. Every Sidekiq process only ever works against one Redis so server-side middleware only needs `redis {|c| PORO.new(c) }`. Your client-side middleware should do `redis_pool.with {|c| PORO.new(c).run }`. If you use `redis` it will work in the default case but will break when sharding.

OK thanks @mperham, I've made those changes  and now we're on the latest versions of both gems, no longer getting the thread errors. 👍
